internal/dag/ingress_processor.go

```diff
@@ -77,34 +82,102 @@ func (p *IngressProcessor) Run(dag *DAG, source *KubernetesCache) {
 // computeSecureVirtualhosts populates tls parameters of
 // secure virtual hosts.
 func (p *IngressProcessor) computeSecureVirtualhosts() {
+	var (
+		fallbackCert *Secret
+		err          error
+	)
+	if p.FallbackCertificate != nil {
+		fallbackCert, err = p.source.LookupSecret(*p.FallbackCertificate, validTLSSecret)
+		if err != nil {
+			p.WithError(err).
+				WithField("secret", p.FallbackCertificate).
+				Error("fallback certificate is invalid")
+		}
+	}
+
 	for _, ing := range p.source.ingresses {
-		for _, tls := range ing.Spec.TLS {
-			secretName := k8s.NamespacedNameFrom(tls.SecretName, k8s.TLSCertAnnotationNamespace(ing), k8s.DefaultNamespace(ing.GetNamespace()))
-			sec, err := p.source.LookupSecret(secretName, validTLSSecret)
-			if err != nil {
-				p.WithError(err).
-					WithField("name", ing.GetName()).
+		fallbackCertEnabled := annotation.FallbackCertificateEnabled(ing)
+		if fallbackCertEnabled {
+			switch {
+			// Case where we have a fallback cert.
+			case fallbackCert != nil:
+				if !p.source.DelegationPermitted(*p.FallbackCertificate, ing.GetNamespace()) {
+					p.WithField("name", ing.GetName()).
+						WithField("namespace", ing.GetNamespace()).
+						WithField("secret", p.FallbackCertificate).
+						Error("fallback certificate delegation not permitted")
+
+					// Disable fallback usage for this Ingress so we can
+					// program the rest of it.
+					fallbackCertEnabled = false
+				}
+
+			// Fallback cert configured but invalid.
+			case p.FallbackCertificate != nil:
+				p.WithField("name", ing.GetName()).
 					WithField("namespace", ing.GetNamespace()).
-					WithField("secret", secretName).
-					Error("unresolved secret reference")
-				continue
-			}
+					Error("fallback certificate enabled but the fallback Certificate Secret is invalid")
+
+				// Disable fallback usage for this Ingress so we can
+				// program the rest of it.
+				fallbackCertEnabled = false
 
-			if !p.source.DelegationPermitted(secretName, ing.GetNamespace()) {
-				p.WithError(err).
-					WithField("name", ing.GetName()).
+			// No fallback cert configured.
+			default:
+				p.WithField("name", ing.GetName()).
 					WithField("namespace", ing.GetNamespace()).
-					WithField("secret", secretName).
-					Error("certificate delegation not permitted")
-				continue
+					Error("fallback certificate enabled but the fallback Certificate Secret is not configured in Contour configuration file")
+
+				// Disable fallback usage for this Ingress so we can
+				// program the rest of it.
+				fallbackCertEnabled = false
+			}
+		}
+
+		for _, tls := range ing.Spec.TLS {
+			var secret *Secret
+			if tls.SecretName == "" {
+				if fallbackCertEnabled {
+					// If no provided secret, use fallback cert as main vhost secret.
+					secret = fallbackCert
```

Comment:
Is this the same behavior as for HTTPProxy? (started looking but if you've already tracked this down I'll spare myself the spelunk).

Reply:
HTTPProxy requires you to specify a TLS cert so you always get a "main" TLS cert for most requests and the fallback for ones w/o SNI

with Ingress you don't have to provide a cert so this solution is kind of stretching of the meaning of "fallback" to extend to that use case